source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRTargetBuilderDirector.mm

```diff
@@ -0,0 +1,268 @@
+                                                      config:director.adaptiveHostConfig
+                                                    rootView:director.rootView
+                                                      button:button];
+        [button addTarget:target
+                      action:@selector(toggleVisibilityOfShowCard)
+            forControlEvents:UIControlEventTouchUpInside];
+        return target;
+    }
+    return nil;
+}
+
+@end
+
+@interface ACRToggleVisibilityTargetBuilder : ACRTargetBuilder
+@end
+
+@implementation ACRToggleVisibilityTargetBuilder
+
++ (ACRToggleVisibilityTargetBuilder *)getInstance {
+    static ACRToggleVisibilityTargetBuilder *singletonInstance = [[self alloc] init];
+    return singletonInstance;
+}
+
+- (NSObject *)build:(std::shared_ptr<BaseActionElement> const &)action
+           director:(ACRTargetBuilderDirector *)director {
+    std::shared_ptr<ToggleVisibilityAction> toggleVisibilityAction =
+        std::dynamic_pointer_cast<ToggleVisibilityAction>(action);
+
+    if (toggleVisibilityAction && director.adaptiveHostConfig && director.rootView) {
+        return [[ACRToggleVisibilityTarget alloc] initWithActionElement:toggleVisibilityAction
+                                                                 config:director.adaptiveHostConfig
+                                                               rootView:director.rootView];
+    }
+    return nil;
+}
+
+- (NSObject *)build:(std::shared_ptr<BaseActionElement> const &)action
+           director:(ACRTargetBuilderDirector *)director
+          ForButton:(UIButton *)button {
+    NSObject *target = [self build:action director:director];
+    if (target) {
+        [button addTarget:target
+                      action:@selector(doSelectAction)
+            forControlEvents:UIControlEventTouchUpInside];
+    }
+    return target;
+}
+
+@end
+
+@implementation ACRTargetBuilderDirector {
+    NSArray<NSDictionary<NSNumber *, ACRTargetBuilder *> *> *_builders;
+}
+
+- (instancetype)init {
+    self = [self init:nil capability:ACRAction adaptiveHostConfig:nil];
+    return self;
+}
+
+- (instancetype)init:(ACRView *)rootView
+            capability:(ACRTargetCapability)capability
+    adaptiveHostConfig:(ACOHostConfig *)adaptiveHostConfig {
+    self = [super init];
+    if (self) {
+        self.builderCapability = capability;
+        self.rootView = rootView;
+        self.adaptiveHostConfig = adaptiveHostConfig;
+
+        NSNumber *openUrl = [NSNumber numberWithInt:static_cast<int>(ActionType::OpenUrl)];
+        NSNumber *submit = [NSNumber numberWithInt:static_cast<int>(ActionType::Submit)];
+        NSNumber *showcard = [NSNumber numberWithInt:static_cast<int>(ActionType::ShowCard)];
+        NSNumber *toggle = [NSNumber numberWithInt:static_cast<int>(ActionType::ToggleVisibility)];
+
+        // target capabilities list supported events and corresponding target builders
+        _builders = @[
+            // ACRAction
```

Comment:
>// ACRAction [](start = 12, length = 12)

Kind of fragile to assume the enum values like this. If someone changes the enum without updating this code, it will continue to build but won't work correctly.

Reply:
There's no need for this to be an array, is there? Each BuilderDirector is used for just one capability, so we only need to keep track of the builders for that one capability, not for all capabilities.

---
In reply to: [321506664](https://github.com/microsoft/AdaptiveCards/pull/3354#discussion_r321506664) [](ancestors = 321506664)

Reply:
I agree with you on the fragility. are you suggesting to have different classes of BuilderDirector? would that be too many different types of directors when their difference is only a capability.

---
In reply to: [321509166](https://github.com/microsoft/AdaptiveCards/pull/3354#discussion_r321509166) [](ancestors = 321509166,321506664)

Reply:
No, not different classes. Just each instance only maintains info for its own set of builders.


```
switch (_builderCapability)
{
    case ACRAction:
        _builders = @{ openUrl : openUrlBuilder, submit : submitBuilder, etc. };
        break;
    case ACRSelectAction:
        _builders = // etc
}
```

---
In reply to: [321511766](https://github.com/microsoft/AdaptiveCards/pull/3354#discussion_r321511766) [](ancestors = 321511766,321509166,321506664)

Reply:
got it thanks

---
In reply to: [321519397](https://github.com/microsoft/AdaptiveCards/pull/3354#discussion_r321519397) [](ancestors = 321519397,321511766,321509166,321506664)